packages/circuits/circuits/proveSubsidyKey.circom

```diff
@@ -0,0 +1,128 @@
+/*
+    Prove:
+        1. if user has a leaf in an existed global state tree
+        2. user state rep < 0
+        3. output all epoch keys for epoch
+*/
+
+include "../node_modules/circomlib/circuits/comparators.circom";
+include "../node_modules/circomlib/circuits/gates.circom";
+include "../node_modules/circomlib/circuits/mux1.circom";
+include "../node_modules/circomlib/circuits/poseidon.circom";
+include "../node_modules/circomlib/circuits/bitify.circom";
+include "../node_modules/circomlib/circuits/sign.circom";
+include "./sparseMerkleTree.circom";
+include "./identityCommitment.circom";
+include "./incrementalMerkleTree.circom";
+include "./modulo.circom";
+
+template ProveSubsidyKey(GST_tree_depth, user_state_tree_depth, epoch_tree_depth, NUM_EPOCH_KEY_NONCE_PER_EPOCH, MAX_REPUTATION_SCORE_BITS) {
+    signal input epoch;
+
+    // Global state tree leaf: Identity & user state root
+    signal private input identity_nullifier;
+    signal private input identity_trapdoor;
+    signal private input user_tree_root;
+    // Global state tree
+    signal private input GST_path_index[GST_tree_depth];
+    signal private input GST_path_elements[GST_tree_depth][1];
+    signal output GST_root;
+    // Attester to prove reputation from
+    signal input attester_id;
+    // Attestation by the attester
+    signal private input pos_rep;
+    signal private input neg_rep;
+    signal private input graffiti;
+    signal private input sign_up;
+    signal private input UST_path_elements[user_state_tree_depth][1];
+    signal input minRep;
+    signal input notEpochKey;
+    signal output subsidyKey;
+    signal output epochKey;
+
+    // we only need to verify that one epk is in the gst
+    // we can then simply calculate the others
+    /* 1. Calculate subsidy key and epoch key */
+
+    component subsidyKeyHasher = Poseidon(2);
+
+    subsidyKeyHasher.inputs[0] <== identity_nullifier + NUM_EPOCH_KEY_NONCE_PER_EPOCH;
```

Comment:
Attestations sent to the `subsidyKey` will not be processed in the unirep protocol
It will not be calculated in the [`processAttestations.circom`](https://github.com/Unirep/Unirep/blob/main/packages/circuits/circuits/processAttestations.circom#L84) and [`userStateTransition.circom`](https://github.com/Unirep/Unirep/blob/a9e808a3435fc14ff9aebb9d2dbb3147168b6b35/packages/circuits/circuits/userStateTransition.circom#L123)

The protocol will process nonce from `0` to `NUM_EPOCH_KEY_NONCE_PER_EPOCH-1`

Reply:
Yes, the subsidyKey is just used for tracking the spent subsidy. The circuit outputs epoch key with nonce 0. I think I could actually remove the subsidy key and just use epoch key 0.